xapian-core/cluster/kmeans.cc

```diff
@@ -61,22 +70,85 @@ KMeans::initialize_centroids(ClusterSet &cset) {
 }
 
 void
-KMeans::initialize_points(MSetDocumentSource source, TermListGroup tlg) {
-    while(!source.at_end()) {
+KMeans::initialize_kmeanspp(ClusterSet &cset) {
```

Comment:
Would it perhaps be better to have a separate function or class to create the initial set of clusters using KMeans++, and a way of initialising `KMeans` with an existing `ClusterSet`? That feels like it's going to flow better into having a PSO implementation, and also allow experimentation with other ways of initialising the clusters in future.